code/app/src/main/java/com/example/bread/repository/ParticipantRepository.java

```diff
@@ -88,12 +101,47 @@ public void addParticipant(@NonNull Participant participant, @NonNull OnSuccessL
                 .addOnFailureListener(onFailureListener != null ? onFailureListener : e -> Log.e("ParticipantRepository", "Failed to add participant: " + participant, e));
     }
 
-    public void addFollower(@NonNull String username, String followerUsername, @NonNull OnSuccessListener<Void> onSuccessListener, OnFailureListener onFailureListener) {
-        // TODO: check if already followed
-        // TODO: decide on what to store in the follower document
-        getParticipantCollRef().document(username).collection("followers").document(followerUsername).set(new Object())
-                .addOnSuccessListener(onSuccessListener)
-                .addOnFailureListener(onFailureListener != null ? onFailureListener : e -> Log.e("ParticipantRepository", "Failed to add follower: " + followerUsername + " to participant: " + username, e));
+    public void createFollowRequest(String senderId, String receiverId, OnSuccessListener<Void> onSuccessListener, OnFailureListener onFailureListener) {
+        if (senderId.equals(receiverId)) {
+            onFailureListener.onFailure(new Exception("Cannot follow oneself."));
+            return;
+        }
+
+        // Check if the follow request already exists
+        db.collection("followRequests")
+                .whereEqualTo("senderId", senderId)
+                .whereEqualTo("receiverId", receiverId)
+                .get()
+                .addOnCompleteListener(task -> {
+                    if (task.isSuccessful() && !task.getResult().isEmpty()) {
+                        onFailureListener.onFailure(new Exception("Follow request already sent."));
+                    } else {
+                        // Create the follow request if not existing
+                        DocumentReference followRequestRef = db.collection("followRequests").document();
```

Comment:
In the document() method, it should explicitly include the followRequestId that was autogenerated; otherwise, Firebase won’t be able to differentiate between requests. I'm not sure if Firebase sets it automatically, but I’d prefer that we set it explicitly.